### DISCUSSION TITLE
Docker run works, docker compose: Network Connection Issue

### DISCUSSION THREAD

**Discussion by adaniels21487** [abraunegg/onedrive#2400] [Q&A] [answered]
Hi Team,
I have been trying to get onedrive up and running in a docker container on a linux based NAS (QNAP) using docker-compose.
I followed the instructions [here](https://github.com/abraunegg/onedrive/blob/master/docs/Docker.md).

When I run the command via docker run, everything works fine:
```
[admin@NAS onedrive]# docker run -it --name onedrive -v "/home/aaron/onedrive/conf:/onedrive/conf"     -v "/home/aaron/onedrive/data:/onedrive/data" -e "ONEDRIVE_UID=1000" -e "ONEDRIVE_GID=1000" -e "ONEDRIVE_VERBOSE=1" -e "ONEDRIVE_DEBUG_HTTPS=1" driveone/onedrive:latest
Base Args: --monitor
# We are being verbose
# Adding --verbose
# We are performing HTTPS debug output
# Adding --debug-https
# Launching onedrive
Using 'user' Config Dir: /onedrive/conf
Configuration file successfully loaded
Checking Application Version ...
Initializing the OneDrive API ...
Configuring Global Azure AD Endpoints
Using Curl defaults for all HTTP operations
Opening the item database ...
All operations will be performed in: /onedrive/data
*   Trying 20.190.142.163:443...
* Connected to login.microsoftonline.com (20.190.142.163) port 443 (#0)
* ALPN, offering h2
^CGot termination signal, performing clean up
Shutting down the HTTP instance
Shutting down db connection and merging temporary data
[admin@NAS onedrive]#
````
All my files are sync'd, UID and GID is correctly set, happy days.

But, when I run the same thing using docker-compose, I get the network connection issue error:
```
[admin@NAS onedrive]# docker-compose up
Creating onedrive_onedrive_1 ... done
Attaching to onedrive_onedrive_1
onedrive_1  | Base Args: --monitor
onedrive_1  | # We are being verbose
onedrive_1  | # Adding --verbose
onedrive_1  | # We are performing HTTPS debug output
onedrive_1  | # Adding --debug-https
onedrive_1  | # Launching onedrive
onedrive_1  | Using 'user' Config Dir: /onedrive/conf
onedrive_1  | Configuration file successfully loaded
onedrive_1  |
onedrive_1  |   Calling Function: testNetwork()
onedrive_1  |
onedrive_1  | Cannot connect to Microsoft OneDrive Service - Network Connection Issue
onedrive_1  | ERROR: Microsoft OneDrive API returned an error with the following message:
onedrive_1  |   Error Message:    Timeout was reached on handle 555CC8E83F70
onedrive_1  | Unable to reach Microsoft OneDrive API service at this point in time, re-trying network tests
onedrive_1  |
onedrive_1  |
onedrive_1  |   Calling Function: testNetwork()
onedrive_1  |
onedrive_1  | Cannot connect to Microsoft OneDrive Service - Network Connection Issue
onedrive_1  | ERROR: Microsoft OneDrive API returned an error with the following message:
onedrive_1  |   Error Message:    Timeout was reached on handle 555CC8E83F70
^CGracefully stopping... (press Ctrl+C again to force)
Stopping onedrive_onedrive_1 ... done
[admin@NAS onedrive]#
```
I downloaded the default user config file from [here](https://raw.githubusercontent.com/abraunegg/onedrive/master/config). Both methods show **Configuration file successfully loaded** so both appear to be reading the same conf and data folders.

I am using the standard docker-compose, with debugging enabled:
```
version: "3"
services:
  onedrive:
    image: driveone/onedrive:latest
    restart: unless-stopped
    environment:
      - ONEDRIVE_UID=1000
      - ONEDRIVE_GID=1000
      - ONEDRIVE_VERBOSE=1
      - ONEDRIVE_DEBUG_HTTPS=1
    volumes:
      - ./conf:/onedrive/conf
      - ./data:/onedrive/data
```

I have followed the guidance in the [Network Connection Issue](https://github.com/abraunegg/onedrive/wiki/Frequently-Asked-Questions#cannot-connect-to-microsoft-onedrive-service---network-connection-issue) section of the FAQ. I can run both curl commands from a busybox container, and as discussed above, connectivity works from the onedrive container, but only when run manually.

This is all on a NAS, so I run several containers from docker-compose and haven't had any issues like this before.
I'm not sure where to look next.

Has anyone seen this before, or have any guidance on how I could further troubleshoot this?

Thanks,
Aaron

**Comment by abraunegg** [comment]
@adaniels21487 
> `driveone/onedrive:latest`

This means you are using a Ubuntu Docker Container.

Because of this, you may also be experiencing this issue due to a misbehaving DNS Servers and/or Firewall that will only give one reply to a parallel A and AAAA request. Whatever DNS Server the Ubuntu Docker Image is using, is what is causing your problem. Additionally, this is most prevalent when your system is using IPv4 and IPv6 network address spaces at the same time.

Essentially a component of your network:
* Ubuntu's Firewall (ufw)
* Your NAS Firewall
* Your Router and/or Firewall that is providing DNS resolution
* You actually have slow DNS resolution - run a local DNS caching server on your home network rather than using your 'default' router DNS configuration (and actually - you should actually do that to ensure you have a better class of DNS resolution anyway)

Any of these above items are potentially causing the issue, and mainly because Ubuntu chose a certain path to resolve DNS resolution issues.

Your options for resolving:
* Dont use a Ubuntu Docker Container - use the Fedora or Alpine image instead as this has not been seen with those images (or reported)
* Disable IPv6 on your Docker ecosystem, forcing the Docker container to only use IPv4.
* Fix your linux based NAS (QNAP) by updating your /etc/resolv.conf on your NAS as per:
![image](https://github.com/abraunegg/onedrive/assets/4956234/c5d25595-fdf8-4423-ad14-10fd78e9f509)
This may help, but im less confident that it will
* If you want yo keep using Ubuntu based Docker container, potentially update the same /etc/resolv.conf change inside the Docker container
* Complain to Ubuntu about their DNS resolver service and get them to fix it
* Build and run your own local caching DNS Server rather than relying on your router and/or other embedded device to provide DNS resolution services.

**Comment by adaniels21487** [comment]
Hi @abraunegg.
Thanks for your in-depth response.
But unfortunately, no luck so far.

I tried using the Alpine and Fedora containers.
I also tried the option single-request in my resolv.conf
IPv6 is disabled on the docker network that the onedrive container uses:
```
[admin@NAS onedrive]# docker inspect 35a062dd5acf
-- SNIP --

        "Name": "/onedrive_onedrive_1",
        "RestartCount": 0,

-- SNIP --

            "NetworkMode": "onedrive_default",

-- SNIP --

        "NetworkSettings": {
            "Bridge": "",
            "SandboxID": "80591a315fb556173151c035b67a8d2e36b3d7a7b82f6a84e7fbf0f67ec6c13e",
            "HairpinMode": false,
            "LinkLocalIPv6Address": "",
            "LinkLocalIPv6PrefixLen": 0,
            "Ports": {},
            "SandboxKey": "/var/run/docker/netns/80591a315fb5",
            "SecondaryIPAddresses": null,
            "SecondaryIPv6Addresses": null,
            "EndpointID": "",
            "Gateway": "",
            "GlobalIPv6Address": "",
            "GlobalIPv6PrefixLen": 0,
            "IPAddress": "",
            "IPPrefixLen": 0,
            "IPv6Gateway": "",
            "MacAddress": "",
            "Networks": {
                "onedrive_default": {
                    "IPAMConfig": null,
                    "Links": null,
                    "Aliases": [
                        "35a062dd5acf",
                        "onedrive"
                    ],
                    "NetworkID": "2fee5be8049559216b9f703bf843b7953ae593e83af28b78d23bdf7c84b59395",
                    "EndpointID": "",
                    "Gateway": "",
                    "IPAddress": "",
                    "IPPrefixLen": 0,
                    "IPv6Gateway": "",
                    "GlobalIPv6Address": "",
                    "GlobalIPv6PrefixLen": 0,
                    "MacAddress": "",
                    "DriverOpts": null
                }
            }
        }
    }
]
[admin@NAS onedrive]#
```
These network settings are the same under both docker run and docker-compose

Still no change, docker run works, but docker-compose doesn't.

**Comment by abraunegg** [comment]
@adaniels21487 
> I tried using the Alpine and Fedora containers.

OK - interesting - so it will be something in the Docker setup then that is preventing correct communication.

Potentially have a read of this as well: https://github.com/abraunegg/onedrive/discussions/2039

**Comment by adaniels21487** [comment, marked as answer]
After much troubleshooting, I wasn't able to find my issue but I was able to resolve it.

I found this excellent description of the network differences between docker run and docker-compose:
https://github.com/moby/moby/issues/41003#issuecomment-631423909

So docker-compose creates a custom network and docker run uses the built-in bridge network.

With this knowledge, I was able to test by making docker run use the docker-compose network by adding:
```
--network onedrive_default
```

Progress, docker run now also fails when using the docker compose network.

In my docker logs I saw:
```
time="2023-05-16T21:05:20.368297832+10:00" level=debug msg="Name To resolve: login.microsoftonline.com."
time="2023-05-16T21:05:20.368532672+10:00" level=debug msg="[resolver] query login.microsoftonline.com. (A) from 172.29.0.2:40196, forwarding to udp:10.0.3.1"
time="2023-05-16T21:05:24.369533624+10:00" level=debug msg="[resolver] read from DNS server failed, read udp 172.29.0.2:40196->10.0.3.1:53: i/o timeout"
time="2023-05-16T21:05:25.373361500+10:00" level=debug msg="Name To resolve: login.microsoftonline.com."
time="2023-05-16T21:05:25.373588624+10:00" level=debug msg="[resolver] query login.microsoftonline.com. (A) from 172.29.0.2:32981, forwarding to udp:10.0.3.1"
time="2023-05-16T21:05:29.373792272+10:00" level=debug msg="[resolver] read from DNS server failed, read udp 172.29.0.2:32981->10.0.3.1:53: i/o timeout"
```

i/o timeout, interesting. Various search results point to this being a host resolver issue.
but all my other docker-compose containers are resolving fine.

So I blow away the docker-compose network:
```
docker network rm onedrive_default
```

and let docker recreate it:
```
docker-compose up
```

and just like that, it's all working. Tested with Ubuntu, Fedora, and Alpine:
```
[administrator@NAS onedrive]$ docker-compose up
Creating network "onedrive_default" with the default driver
Creating onedrive_onedrive_1 ... done
Attaching to onedrive_onedrive_1
onedrive_1  | usermod: no changes
onedrive_1  | Base Args: --monitor
onedrive_1  | # We are being verbose
onedrive_1  | # Adding --verbose
onedrive_1  | # We are performing HTTPS debug output
onedrive_1  | # Adding --debug-https
onedrive_1  | # Launching onedrive
onedrive_1  | Using 'user' Config Dir: /onedrive/conf
onedrive_1  | Configuration file successfully loaded
onedrive_1  | Checking Application Version ...
onedrive_1  | Initializing the OneDrive API ...
onedrive_1  | Configuring Global Azure AD Endpoints
onedrive_1  | Using Curl defaults for all HTTP operations
onedrive_1  | Opening the item database ...
onedrive_1  | All operations will be performed in: /onedrive/data
onedrive_1  | *   Trying 20.190.167.18:443...
onedrive_1  | * Connected to login.microsoftonline.com (20.190.167.18) port 443 (#0)
onedrive_1  | * ALPN, offering h2
onedrive_1  | * ALPN, offering http/1.1
onedrive_1  | *  CAfile: /etc/pki/tls/certs/ca-bundle.crt
```

Thanks @abraunegg for your help.
I hope this helps someone else in the future.
